File: src/bin/dead_ref.cpp
#include <iostream>
#include <vector>

void case1() {
    auto arr = std::vector{1, 2, 3};
    auto &ref = arr[0];
    auto cap = arr.capacity();
    for (auto i = arr.size(); i <= cap; ++i) {
        arr.push_back(i);
    }
    std::cout << "ref: " << ref << std::endl;

    // PANIC!!!

    // Output is not deterministic.
}

void case2() {
    auto arr = std::vector{1, 2, 3};
    auto cap = arr.capacity();
    for (auto i = arr.size(); i <= cap; ++i) {
        arr.push_back(i);
    }
    auto &ref = arr[0];
    std::cout << "ref: " << ref << std::endl;

    // Output:
    // ref: 1
}

int main() {
    case1();
    case2();
    return 0;
}

File: src/bin/dead_ref.rs
// Compile Error:
// cannot borrow `arr` as mutable because it is also borrowed as immutable mutable borrow occurs here
// fn case1() {
//     let mut arr = vec![1, 2, 3];
//     let cap = arr.capacity();
//     let mut iter = arr.iter();
//     for _ in arr.len()..=cap {
//         arr.push(0);
//     }
//     println!("first: {}", iter.next().unwrap());
// }
fn case2() {
    let mut arr = vec![1, 2, 3];
    let cap = arr.capacity();
    for _ in arr.len()..=cap {
        arr.push(0);
    }
    let mut iter = arr.iter();
    println!("first: {}", iter.next().unwrap());

    // Output:
    // first: 1
}
fn main() {
    // case1();
    case2();
}
